src/python/pants/init/plugin_resolver.py

```diff
@@ -1,201 +1,140 @@
 # Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-import hashlib
 import logging
-import os
-import shutil
 import site
-import uuid
-from typing import Iterable, Iterator, List, Optional, Type, TypeVar, cast
+import sys
+from typing import Optional, TypeVar, cast
 
-from pex import resolver
-from pex.interpreter import PythonInterpreter
-from pex.network_configuration import NetworkConfiguration
-from pkg_resources import Distribution, WorkingSet
+from pkg_resources import WorkingSet
 from pkg_resources import working_set as global_working_set
 
+from pants.backend.python.util_rules.pex import (
+    PexInterpreterConstraints,
+    PexRequest,
+    PexRequirements,
+    VenvPex,
+    VenvPexProcess,
+)
+from pants.core.util_rules.pants_environment import PantsEnvironment
+from pants.engine.collection import DeduplicatedCollection
+from pants.engine.internals.session import SessionValues
+from pants.engine.process import ProcessCacheScope, ProcessResult
+from pants.engine.rules import Get, QueryRule, collect_rules, rule
+from pants.init.bootstrap_scheduler import BootstrapScheduler
 from pants.option.global_options import GlobalOptions
-from pants.option.optionable import Optionable
 from pants.option.options_bootstrapper import OptionsBootstrapper
-from pants.option.scope import ScopeInfo
 from pants.option.subsystem import Subsystem
-from pants.python.python_repos import PythonRepos
-from pants.util.contextutil import temporary_dir
-from pants.util.dirutil import safe_concurrent_rename, safe_delete, safe_open
-from pants.util.memo import memoized_property
-from pants.version import PANTS_SEMVER
+from pants.util.logging import LogLevel
 
 logger = logging.getLogger(__name__)
 
 
 S = TypeVar("S", bound=Subsystem)
 
 
-class PluginResolver:
-    @staticmethod
-    def _is_wheel(path) -> bool:
-        return os.path.isfile(path) and path.endswith(".whl")
+class ResolvedPluginDistributions(DeduplicatedCollection[str]):
+    sort_input = True
+
+
+@rule
+async def resolve_plugins(
+    interpreter_constraints: PexInterpreterConstraints, global_options: GlobalOptions
```

Comment:
It looks like we solely use the interpreter_constraints for the sake of testing? Could we remove that to simplify this?

If we need to keep it, how about we newtype? I think we probably want to avoid introducing `PexInterpreterConstraints` as a param to be used in rules, as it's very generic and vague what it refers to - we want to namespace the domain it's being used for.

Reply:
I think that it is worth keeping for the purposes of validating that these resolves are in fact interpreter specific.

Reply:
Too late with the merge, but how about newtyping it?